core/vm/environment.go

```diff
@@ -17,110 +17,288 @@
+	// only.
+	contract := NewContract(caller, to, value, gas)
+	contract.SetCallCode(&addr, env.StateDB.GetCodeHash(addr), env.StateDB.GetCode(addr))
+	defer contract.Finalise()
+
+	ret, err := env.EVM().Run(contract, input)
+	if err != nil {
+		contract.UseGas(contract.Gas)
+
+		env.StateDB.RevertToSnapshot(snapshotPreTransfer)
+	}
+
+	return ret, err
 }
 
-// Database is a EVM database for full state querying.
-type Database interface {
-	GetAccount(common.Address) Account
-	CreateAccount(common.Address) Account
+// Same as CallCode except sender and value is propagated from parent to child scope
+func (env *Environment) DelegateCall(caller ContractRef, addr common.Address, input []byte, gas *big.Int) ([]byte, error) {
+	if env.vmConfig.NoRecursion && env.Depth > 0 {
+		caller.ReturnGas(gas)
 
-	AddBalance(common.Address, *big.Int)
-	GetBalance(common.Address) *big.Int
+		return nil, nil
+	}
 
-	GetNonce(common.Address) uint64
-	SetNonce(common.Address, uint64)
+	// Depth check execution. Fail if we're trying to execute above the
+	// limit.
+	if env.Depth > int(params.CallCreateDepth.Int64()) {
+		caller.ReturnGas(gas)
+		return nil, DepthError
+	}
 
-	GetCodeHash(common.Address) common.Hash
-	GetCodeSize(common.Address) int
-	GetCode(common.Address) []byte
-	SetCode(common.Address, []byte)
+	var (
+		snapshot = env.StateDB.Snapshot()
+		to       = env.StateDB.GetAccount(caller.Address())
+	)
 
-	AddRefund(*big.Int)
-	GetRefund() *big.Int
+	// Iinitialise a new contract and make initialise the delegate values
+	contract := NewContract(caller, to, caller.Value(), gas).AsDelegate()
+	contract.SetCallCode(&addr, env.StateDB.GetCodeHash(addr), env.StateDB.GetCode(addr))
+	defer contract.Finalise()
 
-	GetState(common.Address, common.Hash) common.Hash
-	SetState(common.Address, common.Hash, common.Hash)
+	ret, err := env.EVM().Run(contract, input)
+	if err != nil {
+		contract.UseGas(contract.Gas)
 
-	Suicide(common.Address) bool
-	HasSuicided(common.Address) bool
+		env.StateDB.RevertToSnapshot(snapshot)
+	}
 
-	// Exist reports whether the given account exists in state.
-	// Notably this should also return true for suicided accounts.
-	Exist(common.Address) bool
-	// Empty returns whether the given account is empty. Empty
-	// is defined according to EIP161 (balance = nonce = code = 0).
-	Empty(common.Address) bool
+	return ret, err
 }
 
-// Account represents a contract or basic ethereum account.
-type Account interface {
-	SubBalance(amount *big.Int)
-	AddBalance(amount *big.Int)
-	SetBalance(*big.Int)
-	SetNonce(uint64)
-	Balance() *big.Int
-	Address() common.Address
-	ReturnGas(*big.Int, *big.Int)
-	SetCode(common.Hash, []byte)
-	ForEachStorage(cb func(key, value common.Hash) bool)
-	Value() *big.Int
+// Create a new contract
+func (env *Environment) Create(caller ContractRef, code []byte, gas, value *big.Int) ([]byte, common.Address, error) {
+	if env.vmConfig.NoRecursion && env.Depth > 0 {
```

Comment:
Could you please add a comment that this is done for testing and doesn't execute during normal VM operation?